src/queue.c

```diff
@@ -3992,16 +4037,21 @@ _dispatch_global_queue_poke_slow(dispatch_queue_t dq, unsigned int n)
 			}
 		}
 	}
-	uint32_t j, t_count;
+	int32_t j, t_count;
 	// seq_cst with atomic store to tail <rdar://problem/16932833>
 	t_count = os_atomic_load2o(qc, dgq_thread_pool_size, ordered);
 	do {
-		if (!t_count) {
+#if HAVE_INTERNAL_PTHREAD_WORKQUEUE
+		bool pool_full = t_count <= 0; // oversubscription encoded by negative dgq_thread_pool_size
+#else
+		bool pool_full = t_count == 0;
+#endif
+		if (pool_full) {
 			_dispatch_root_queue_debug("pthread pool is full for root queue: "
 					"%p", dq);
 			return;
 		}
-		j = i > t_count ? t_count : i;
+		j = (int32_t)i > t_count ? t_count : (int32_t)i;
```

Comment:
please avoid this cast by making `i` signed, which will mean fixing `_dispatch_global_queue_poke` to take an int too, all the way down to the root of that need which is `da_thr_cnt`. this is fine as dispatch really only can manage in the thousands of threads, so making this signed is not an issue, and that would avoid these ugly looking casts